### PR TITLE
Run golangci-lint in CI, address linter errors

### PR DESCRIPTION
**Pull request by ti-mo**
@lmb As discussed. I'd like us to consider adding gocyclo, prealloc, gosec and maligned later, since some of them flagged a few things that I didn't understand at first glance, and some are more invasive.

PTAL!

### CODE REVIEW

**Review by lmb — CHANGES_REQUESTED** [review]
These are nice fixes, thanks!

What do you want to get out of adding the linters? I would like them to have no or very low false positives and to identify issues that will (most likely) trigger bugs. Otherwise they very quickly become a liability instead of an asset, since every additional linter is another hoop that any contributor has to jump through.

Without further ado:
* misspell: let's drop this. I'll take a typo over a failed CI run.
* structcheck: too many false positives. It's nice that we can disable the lint per file but it's still a false positive.
* gocyclo: I'm very wary of tooling that turns a heuristic into a dogma. Sometimes a function is complex, if it's too much we as humans / maintainers can step in and try to clean up.
* prealloc and maligned: let's fix issues like these when there is a perf problem in a real use case, not when the linter yells.
* gosec: a colleague of mine ran this on an internal tool and it created a ton of false positives, hopefully you'll have better luck.

File: .golangci.yaml
Comment (ti-mo) [reply .golangci.yaml:10]: Okay, this might help keeping runs stable across golangci-lint upgrades. :+1:

File: map.go
Comment (ti-mo) [reply map.go:937]: Thanks! Having `_ = (*value).Close()` here communicates that intention to the reader.

File: perf/ring.go
Comment (ti-mo) [reply perf/ring.go:82]: I can't find the reason for `Close()` being ignored by errcheck in either the errcheck or the golangci-lint source, will try to find out why that is. There are a couple of threads complaining about errcheck forcing deferred Close()s to be error-checked as well, but this is not the same.

So yes, the sole reason why I didn't add a check on Close() as well is because the linter seems to be okay with it. Should we make `(perfEventRing).Close()` return error, or is it really fine if Close and Munmap return errors?

**Comment by ti-mo** [comment]
> What do you want to get out of adding the linters?

My editor runs golangci-lint on-save for the whole package. In doing this, the repo-wide `.golangci.yaml` is used, so it's pretty much guaranteed to be consistent with whatever linter runs in CI. This kind of addresses your comment `misspell: let's drop this. I'll take a typo over a failed CI run`. I would very much like to have misspell enabled to avoid committing language blunders or typos in general. (we have an internal Slack channel dedicated to these :joy:)

> I would like them to have no or very low false positives and to identify issues that will (most likely) trigger bugs.

Well, it also depends on what you find worth it as an individual. Some of these tools perform static analysis, some are purely about coding style. At the end of the day, we just need to agree on a config, and that's what the codebase will adhere to from that point on. I personally feel more comfortable adding a few `//nolint`s to make an existing codebase compliant (a few minutes of work, once) if it means I'll get instant feedback on some aspects of code quality for free from that point on. As a recent example, I left some internal struct fields unused refactoring a PR and you had to call that out during review. This is something structcheck will prevent, saving time for both the reviewer and the contributor, and reducing friction. IMO none of the tools I've enabled in this PR are invasive nor are they unnecessarily noisy.

> Otherwise they very quickly become a liability instead of an asset, since every additional linter is another hoop that any contributor has to jump through.

I follow you, but let's not overblow this please. It looks to me like golangci-lint has become the golden standard; it's a single tool, it's trivial to set up, it just works, and I'd argue this should be part of the default devkit. No hoops, it just uses the repo's linter config by default. If the few linters we have enabled do start causing pain for some reason, they can be disabled.

`gocyclo`: agree on the heuristic argument, but I personally have found value in using this to find long functions in existing code bases. iirc this only flagged a single function in this lib, so enabling this would be low-friction.

`prealloc` and `maligned`: agree with your rationale, but I find it hard to justify leaving alloc efficiency on the table if running a tool can point out exactly what to fix. After the (small) initial effort, it's also set & forget.

`gosec`: Ack, not sure if this will end up being useful, but left it on the list as reminder to explore.

File: internal/btf/btf_types.go
Comment (ti-mo) [reply internal/btf/btf_types.go:39]: Commented this out instead and documented why it is unused.

File: map.go
Comment (ti-mo) [reply map.go:937]: Still not sure if we should ignore *all* errors here, but not an issue for now.

File: perf/ring.go
Comment (ti-mo) [reply perf/ring.go:83]: Opted not to break `ring.Close()` API, if this proves an issue later, we could return error.

**Comment by lmb** [comment]
> My editor runs golangci-lint on-save for the whole package. In doing this, the repo-wide .golangci.yaml is used, so it's pretty much guaranteed to be consistent with whatever linter runs in CI.

But that won't be true in everybody else's environment, for example I don't have golangci-lint installed myself. You can argue that this is bad form, but I won't be the only contributor like that. Now I have two options: install golangci-lint myself and set it up appropriately including editor config and what not, or make a PR and wait for CI to fail. Neither are great imo.

> I follow you, but let's not overblow this please. It looks to me like golangci-lint has become the golden standard; it's a single tool, it's trivial to set up, it just works, and I'd argue this should be part of the default devkit.

OK, let's try the current config and see what happens.

Regarding gocyclo, prealloc and maligned: I have very very big reservations and will push back on enforcing them via CI / blindly changing code to placate them. There are always things that could use a rewrite or aren't as fast as they could be. Let's worry about them when there are bugs or real perf problems.

**Review by lmb — APPROVED** [review]

**Comment by ti-mo** [comment]
> There are always things that could use a rewrite or aren't as fast as they could be. Let's worry about them when there are bugs or real perf problems.

I agree, let's see what they come up with later :+1: Thanks for the review!

**Pull request merged**
